travis-ci: run tests using docker

grondo opened the pull request:
Ok, finally have a working docker integration with travis, built off @trws Dockerfiles and initial work in this area. The changes of note from Tom's work include:

 - Combine gcc (default), gcc 8, and clang 6.0 into a single Ubuntu 18.04 (bionic) image (it was just easier for me to work with one "base" ubuntu image)
 - Add missing deps to Dockerfiles to satisfy all tests
 - Since flux-security will soon be a moving target, remove its build from the "base" image(s), install a "travis" Dockerfile that builds a specified version of flux-security and optionally installs a user matching the userid running `docker build`. This is done so that we don't have to chown the working directory, which then becomes unreadable by the testing user after the build.
 - Added a `src/test/docker/docker-run-checks.sh` which abstracts the build of the ephemeral travis container and execution of the underlying `travis_run.sh` script inside the container. This makes it easy to run the same build as travis from your docker-capable workstation/laptop with a single command. E.g. you can run the default checks with `make -j 8`  in the centos7-base image with:
`src/test/docker/docker-run-checks.sh -j 8 -i centos7-base -- --with-flux-security`. 
```
Usage: docker-run-checks.sh [OPTIONS] -- [CONFIGURE_ARGS...]
Build docker image for travis builds, then run tests inside the new
container as the current user and group.

Uses the current git repo for the build.

Options:
 -h, --help                    Display this message
     --no-cache                Disable docker caching
 -i, --image=NAME              Use base docker image NAME (default=bionic-base)
 -S, --flux-security-version=N Install flux-security vers N (default=0.2.0)

 -j, --jobs=N                  Value for make -j (default=2)

 -d, --distcheck               Run 'make distcheck' instead of 'make check'
 -I, --interactive             Instead of running travis build, run docker
                                image with interactive shell.
```

 - Added a config.site to the `centos7-base` to give a hint to `configure` (for flux-security) that libdir=/usr/lib64 not /usr/lib.
 - redo and simplify `.travis.yml` to reduce number of builds and add CentOS7 build, etc.
 - a few other fixes found along the way.


Before this can go in, unfortunately, we'll need a strategy for flux-sched. We could copy a similar scheme and place a "travis" Dockerfile that builds and installs `flux-security` and `flux-core` and uses a similar scheme as here (maybe the same script)

A better idea might be to have flux-core travis builds automatically upload a docker image with flux-core installed to a `fluxrm/flux-core:latest` tag on Docker Hub, which would simplify not only builds against flux-core but would allow people to easily try out flux-core master by pulling from Docker Hub. We could also use this created tagged Docker images like `fluxrm/flux-core:0.11.0`.

However, I haven't implemented that yet.

Fixes #1648.

grondo commented:
Oh, forgot to mention that currently both `cppcheck` and `pylint` are disabled here because the newer versions in Ubuntu 18.04 generated a lot of errors.

I found that if I let cppcheck parse the system include files, I got a lot less false positives (i.e. with `-I /usr/include`), however, the run took about 10x longer.

trws commented:
This sounds great @grondo!  As to your last point, I was thinking we could split the base images (unfortunately, it would work not to but the result would be *huge*) into a version with run dependencies and a version with build deps, then use the [multi-stage build method](https://docs.docker.com/develop/develop-images/multistage-build/#before-multi-stage-builds) from the docs to pull the built flux out of the build container into the run container and have the travis deploy target (or the script after success, doesn't really matter) push it up to dockerhub with a tag matching the release.  

Doing a version of this that just builds off of the build image should be pretty straightforward, it's the optimizing that would likely be annoying.

grondo commented:
Great suggestions @trws. I might try to punt on the optimization for now and try to get something that flux-sched builder could use, which would be built from the base image. It may be large, but a lot of the stuff needed by the flux-core build is also required for flux-sched, so we could consider this just a "build" image, and optimize the "run" image later...

trws commented:
That’s a good point. We’d need to re-install almost all the build dependencies for sched anyway. Sounds like a good plan to me.

On Sep 21, 2018, at 5:33 PM, Mark Grondona <notifications@github.com<mailto:notifications@github.com>> wrote:


Great suggestions @trws<https://github.com/trws>. I might try to punt on the optimization for now and try to get something that flux-sched builder could use, which would be built from the base image. It may be large, but a lot of the stuff needed by the flux-core build is also required for flux-sched, so we could consider this just a "build" image, and optimize the "run" image later...

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/flux-framework/flux-core/pull/1670#issuecomment-423575484>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAoStYH2TmkDlf4j9WDbRxVwlFI6XEvwks5udQaxgaJpZM4WzA8v>.

grondo commented:
Ok, latest changes here should be very close to ready for merging, including:
 
 * Added fixes for issues causing failures in docker builds, including #1679 and #1674 
 * Optionally "tag" the resulting docker image after `make install` supon successful checks (docker-run-check.sh --tag=NAME) option, and push the result to docker hub. The docker hub image will be named `fluxrm/flux-core:${IMG}-${tag}` where tag is one of `latest` (if built as a result of a merge to master) `vx.y.z` (if built as a result of a tag) and IMG is the name of the image, e.g. `bionic-base`. For bionic-base images, a convenience copy of the tag is created without the `bionic-base-` prefix. E.g. that means there will be a `fluxrm/flux-core:latest` that points to the latest build from master.

This means that `flux-sched`'s build could be modified to pull from `fluxrm/flux-core` and it will get an image with the latest flux-core build installed. If software is tied to a specific version of flux-core, this verison could be pulled down via `fluxrm/flux-core:v0.10.0` (though that doesn't exist yet since it was built before these changes)

The work to take advantage of docker builds in flux-sched has not been done yet.

Finally, you can rerun builds similar to the Travis-CI environment via the `docker-run-checks.sh` script on any docker-enabled machine. For example, to check the centos7 build on an Ubuntu machine:

```
$ src/test/docker/docker-run-checks.sh -j 12 -i centos7-base -- --with-flux-security --enable-caliper
```

Note that the build occurs in your current working directory (i.e. the current git repo is mounted as a volume in the container), and the script will re-run autogen and `make clean` on you. Note also the `--` used to separate the options for the `docker-run-checks.sh` script and the options that are passed to the underlying `./configure`.

To instead run interactively in the image, use the `-I, --interactive` option to the script.

```
 src/test/docker/docker-run-checks.sh -q -I -i centos7-base 
```

garlick commented:
This is pretty cool!

Small nit:  The "update dockerfiles" commit drops pmix from the bionic Docerfile but not the centos7 one.

garlick commented:
Suggestion:  split cmd/flux commits out to their own PR?

grondo commented:
> Small nit: The "update dockerfiles" commit drops pmix from the bionic Dockerfile but not the centos7 one.

Good catch! Thanks.

> Suggestion: split cmd/flux commits out to their own PR?

Yeah, that seems to make logical sense now that I think about it. I'll put those 3 commits into their own PR.

grondo commented:
Ok, I've rebased on master, squashed down the incremental Dockerfile development, and removed pmix from the centos7 image (and updated the centos7-base image on Docker Hub)

I'm still not sure why the coverage dropped, but I can continue to look at that in the background.

grondo commented:
Oh, also the travis-dep-builder.sh script isn't removed yet since it will still be used by flux-sched builds until we can transition to using the `fluxrm/flux-core` docker image for builds.

grondo commented:
Oops forgot to have the centos7 image issue a deploy to Docker Hub. This could wait, as it remains to be seen if the deploy will even work...

garlick commented:
OK, want to merge then?

grondo commented:
Sure. Then we can see if the docker deploy works for Ubuntu

grondo commented:
Eh, darnit. Left the wrong repo slug in the .travis.yml for pushing to docker hub :-(

I'll have to make another PR to fix that issue and will add DOCKER_TAG to the CentOS image so a version that gets deployed (hopefully) as well.

SteVwonder commented:
Wooo! Great work @grondo and @trws!

SteVwonder commented on src/test/docker/centos7-base/Dockerfile:
```diff
@@ -0,0 +1,70 @@
+ && yum clean all
+
+# The cmake from yum is incredibly ancient, download a less ancient one
+RUN wget -q --no-check-certificate https://cmake.org/files/v3.10/cmake-3.10.1-Linux-x86_64.tar.gz\
```
Why no checking of the certificate?  I presume this is more for compatibility when building at LLNL than when building on Travis.

grondo replied:
I'm not sure, but your supposition sounds correct.

trws commented:
Looks great @grondo!
